schema/pgsql/upgrades/025.sql

```diff
@@ -0,0 +1,10 @@
+ALTER TABLE incident ALTER COLUMN severity SET DEFAULT 'ok';
+ALTER TABLE incident_contact ALTER COLUMN role SET DEFAULT 'recipient';
+ALTER TABLE incident_history ALTER COLUMN type SET DEFAULT 'opened';
```

Comment:
The discussion you linked was about integer value ranges. If you use different databases, there will inherently be differences in the behavior. I really don't see why we would want to remove useful checks that are only supported by one database just to be able to say that now all behave poorly. You also added `CHECK` constraints even though they are ignored by some of the [MySQL](https://dev.mysql.com/doc/refman/8.0/en/create-table-check-constraints.html)/[MariaDB](https://mariadb.com/kb/en/constraint/#check-constraints) versions you claim to support:

> Prior to MySQL 8.0.16, CREATE TABLE permits only the following limited version of table CHECK constraint syntax, which is parsed and ignored:
> ```
> CHECK (expr)
> ```

> Constraints are enforced. Before MariaDB 10.2.1 constraint expressions were accepted in the syntax but ignored.

Reply:
> Really, CHECKing only on one RDBMS is pointless.

No, it's not. The check probably represents an assumption made somewhere in the code. With it being checked somewhere, you have a higher chance of noticing earlier during development if you accidentally violate that assumption.

Reply:
I think the context of https://github.com/Icinga/icingadb/pull/221#discussion_r756088881 was quite different than this situation:

- It was about the range of allowed integer values in a column. Here it's about intentionally disabling sound input validation in PostgreSQL just because MySQL/MariaDB can't handle it properly.
- For Icinga DB, we ported from MySQL/MariaDB to PostgreSQL, so there you had the additional argument that existing code could possibly rely on the then already present behavior. We're porting the other way around, so all previous code was developed while the restriction was already in place.

Also, if you want to harmonize the behavior of the two implementations, the goal should be to push MySQL/MariaDB towards the sane behavior, not PostgreSQL towards the strange behavior. That might look something like that, but even that looks strange and I don't think it's necessary, but at least it would improve something overall:

```sql
MariaDB [jbrost]> create table enum_check (x enum('a', 'b'), check (x is not null));
Query OK, 0 rows affected (0.027 sec)

MariaDB [jbrost]> insert into enum_check () values ();
ERROR 4025 (23000): CONSTRAINT `CONSTRAINT_1` failed for `jbrost`.`enum_check`
```

Reply:
Oh, not only the enums have magic defaults. 🙈

```
Server version: 5.5.62 MySQL Community Server (GPL)

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MySQL [idb]> insert into rotation() values ();
Query OK, 1 row affected, 6 warnings (0,007 sec)

MySQL [idb]> select * from rotation;
+----+-------------+----------+------+------+---------+---------------+----------------+
| id | schedule_id | priority | name | mode | options | first_handoff | actual_handoff |
+----+-------------+----------+------+------+---------+---------------+----------------+
|  1 |           0 |        0 |      | 24-7 |         | 0000-00-00    |              0 |
+----+-------------+----------+------+------+---------+---------------+----------------+
1 row in set (0,002 sec)
```

Reply:
Your approach looks nice, but I'm afraid I'll have to make enums like ('NULL', 'a', 'b') for this to work:

```
Server version: 9.0.0 MySQL Community Server - GPL

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MySQL [idb]> insert into schedule(name) values ('');
Query OK, 1 row affected (0,009 sec)

MySQL [idb]> alter table rotation add check(mode is not null);
Query OK, 0 rows affected (0,094 sec)
Records: 0  Duplicates: 0  Warnings: 0

MySQL [idb]> insert into rotation(schedule_id,priority,name,options,first_handoff,actual_handoff) values (1,0,'','','2024-01-01',0);
Query OK, 1 row affected (0,040 sec)

MySQL [idb]> select * from rotation;
+----+-------------+----------+------+------+---------+---------------+----------------+
| id | schedule_id | priority | name | mode | options | first_handoff | actual_handoff |
+----+-------------+----------+------+------+---------+---------------+----------------+
|  1 |           1 |        0 |      | 24-7 |         | 2024-01-01    |              0 |
+----+-------------+----------+------+------+---------+---------------+----------------+
1 row in set (0,002 sec)
```

Shall I?

Reply:
> Your approach looks nice, but I'm afraid I'll have to make enums like ('NULL', 'a', 'b') for this to work:

Why? Your output misses the `create table` statement used. The column must not have a `not null` clause inside the column definition, otherwise MySQL would still apply the default before the dedicated `check` constraint.

> Shall I?

No, adding a fake `'NULL'` value looks even worse. Just set no default, you're creating a problem where none exists.

Reply:
> Sorry, overseen that. Your approach is working. But it allows NULL in pre-8.0.16. 🙈 Is this acceptable?

In my opinion, yes. These old versions will go away over time and that problem will solve itself.

A comment that saying that (a) this is a workaround for MySQL/MariaDB implicitly using default values and (b) the constraint can't be relied on old versions would be helpful though.

Reply:
Actually (b) applies to **all** CHECKs.. 😅